### PR TITLE
Optimise Siddhi MongoDB store query execution

### PR DESCRIPTION
**Pull request by PrabodDunuwila**
## Purpose
- Optimize MongoDB store to execute select clause in DB by pushing the querying to DB level rather than fetch the records to Siddhi level and executing queries at Siddhi level.

## Goals
- Selection of stream, store and constants and supports logical operators, inline math operators, comparison operators and 'is null'. 
- Group by with functions min, max, avg, sum and count. 
- Having, order by, limit and offset.

## Approach
- Implement compileSelection method and query method in MongoDBEventTable. And used a visitor for selection named 'MongoSelectExpressionVisitor'.

## Release note
Optimise Siddhi MongoDB store query execution

## Documentation
N/A as this is optimisation

## Automation tests
 - Unit tests - 91%
 - Integration tests - N/A

## Security checks
- Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
- Ran FindSecurityBugs plugin and verified report? yes
- Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Test environment
JDK 1.8, MongoDB 3.6, 4.2

### CODE REVIEW

**Comment by niveathika** [comment]
@PrabodDunuwila It seems your commit messages are clear enough, but better follow standards when coming up with commit messages, as I see some common flows as not using Imperative mood. 
Follow them from here on out, https://chris.beams.io/posts/git-commit/

**Comment by niveathika** [comment]
Let's add more information to the PR since it is referred from the release note and the above info is not enough. Try not to remove content heading in PR template when it is major improvement/new feature

File: .gitignore [hunk 1]
```diff
@@ -20,3 +20,11 @@
...
 # virtual machine crash logs, see http://www.java.com/en/download/help/error_hotspot.xml
 hs_err_pid*
+
+#idea files
+.idea/
+*.iml
+
+target/
+
+docs/api/2.0.4-SNAPSHOT.md
```
Comment (niveathika) [inline .gitignore]: Do not remove a single file in gitignore. Also, since the auto doc deployment will also use this let's remove this ignore rule.

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: The license year is added/changed only when a file is newly added. Revert this change

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 3]
```diff
@@ -623,4 +626,262 @@ protected void destroy() {
...
+        List<String> attributeList = new ArrayList<>();
+        Document findFilter = MongoTableUtils
+                .resolveCondition((MongoCompiledCondition) compiledCondition, parameterMap);
+        String selectQuery = ((MongoDBCompileSelection) compiledSelection).getCompileSelectQuery();
+        String groupByQuery = ((MongoDBCompileSelection) compiledSelection).getGroupBy();
+        String havingQuery = ((MongoDBCompileSelection) compiledSelection).getHaving();
+        String orderByQuery = ((MongoDBCompileSelection) compiledSelection).getOrderBy();
+        Long limitValue = ((MongoDBCompileSelection) compiledSelection).getLimit();
+        Long offsetValue = ((MongoDBCompileSelection) compiledSelection).getOffset();
+        if (parameterMap.values().size() > 0) {
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Check to see if u can resuse MongoTableUtils.resolveCondition()

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 4]
```diff
@@ -623,4 +626,262 @@ protected void destroy() {
...
+        }
+        if (groupByQuery != null) {
+            Document groupBy = Document.parse(groupByQuery);
+            aggregateList.add(groupBy);
+            MongoTableUtils.printDebugLogs("GroupBy query : ", groupBy.toJson());
+        }
+        if (selectQuery != null) {
+            Document project = Document.parse(selectQuery);
+            aggregateList.add(project);
+            MongoTableUtils.printDebugLogs("Select query : ", project.toJson());
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]:
```suggestion
            MongoTableUtils.logQuery("Select query : ", project.toJson());
```

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 5]
```diff
@@ -623,4 +626,262 @@ protected void destroy() {
...
+
+    @Override
+    protected RecordIterator<Object[]> query(Map<String, Object> parameterMap, CompiledCondition compiledCondition,
+                                             CompiledSelection compiledSelection, Attribute[] outputAttributes)
+            throws ConnectionUnavailableException {
+        List<Document> aggregateList = new ArrayList<>();
+        List<String> attributeList = new ArrayList<>();
+        Document findFilter = MongoTableUtils
+                .resolveCondition((MongoCompiledCondition) compiledCondition, parameterMap);
+        String selectQuery = ((MongoDBCompileSelection) compiledSelection).getCompileSelectQuery();
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Case the compiledSelection variable once, and add local variables only when you need them.

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 6]
```diff
@@ -623,4 +626,262 @@ protected void destroy() {
...
+    }
+
+    private String getProjectionString(List<SelectAttributeBuilder> selectAttributeBuilders) {
+        List<MongoSelectExpressionVisitor> selectExpressionVisitorList =
+                getSelectAttributesList(selectAttributeBuilders);
+        StringBuilder compiledProjectionJSON = new StringBuilder();
+        compiledProjectionJSON.append("{$project:{_id:0, ");
+        for (int i = 0; i < selectExpressionVisitorList.size(); i++) {
+            MongoSelectExpressionVisitor selectExpressionVisitor = selectExpressionVisitorList.get(i);
+            String rename = selectAttributeBuilders.get(i).getRename();
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java:730]: Have you checked rename is null scenario?

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 7]
```diff
@@ -623,4 +626,262 @@ protected void destroy() {
...
+            if (selectExpressionVisitor.getStreamVarCount() == 0 && selectExpressionVisitor.getConstantCount() == 0) {
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: It seems MongoSelectExpressionVisitor always returns these values as 0. Do a check

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 8]
```diff
@@ -623,4 +626,262 @@ protected void destroy() {
...
+
+    private String getGroupByProjectionString(List<SelectAttributeBuilder> selectAttributeBuilders,
+                                              List<ExpressionBuilder> groupByExpressionBuilders) {
+        StringBuilder compiledProjectionJSON = new StringBuilder();
+        List<MongoSelectExpressionVisitor> getSelectExpressionVisitorList =
+                getSelectAttributesList(selectAttributeBuilders);
+        compiledProjectionJSON.append("{$project:{_id:0,");
+        for (int i = 0; i < getSelectExpressionVisitorList.size(); i++) {
+            MongoSelectExpressionVisitor visitor = getSelectExpressionVisitorList.get(i);
+            String rename = selectAttributeBuilders.get(i).getRename();
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java:805]: Add test cases where rename is null

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 9]
```diff
@@ -623,4 +626,262 @@ protected void destroy() {
...
+            }
+        }
+        compiledGroupByJSON.append((groupByAttributesList.size() == 1) ? ',' : "},");
+        List<MongoSelectExpressionVisitor> getSelectExpressionVisitorList =
+                getSelectAttributesList(selectAttributeBuilders);
+        for (int i = 0; i < getSelectExpressionVisitorList.size(); i++) {
+            MongoSelectExpressionVisitor visitor = getSelectExpressionVisitorList.get(i);
+            String rename = selectAttributeBuilders.get(i).getRename();
+            if (!groupByAttributesList.contains(rename)) {
+                if (visitor.getStreamVarCount() == 0 && visitor.getConstantCount() == 0) {
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: This will be always true

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 10]
```diff
@@ -623,4 +626,262 @@ protected void destroy() {
...
+                    String compiledCondition = visitor.getCompiledCondition();
+                    boolean isFunctionUsed = visitor.checkFunctionsUsed();
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]:
```suggestion
                    boolean isFunctionUsed = visitor.isFunctionsPresent();
```

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 11]
```diff
@@ -623,4 +626,262 @@ protected void destroy() {
...
+        }
+        compiledProjectionJSON.append("}");
+        return compiledProjectionJSON.toString();
+    }
+
+    private String getHavingString(ExpressionBuilder havingExpressionBuilder) {
+        MongoExpressionVisitor visitor = new MongoExpressionVisitor();
+        havingExpressionBuilder.build(visitor);
+        String having = visitor.getHavingCompiledCondition();
+        return "{$match:" + having + "}";
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: No need for having variable
```suggestion
        return "{$match:" + visitor.getHavingCompiledCondition() + "}";
```

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 12]
```diff
@@ -623,4 +626,262 @@ protected void destroy() {
...
+    }
+
+    private String getOrderByString(List<OrderByAttributeBuilder> orderByAttributeBuilders) {
+        StringBuilder compiledOrderByJSON = new StringBuilder();
+        List<MongoExpressionVisitor> orderByExpressionVisitorList =
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java:828]: Similar to getGroupByVisistorList - reuse

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 13]
```diff
@@ -623,4 +626,262 @@ protected void destroy() {
...
+        for (int i = 0; i < orderByExpressionVisitorList.size(); i++) {
+            MongoExpressionVisitor visitor = orderByExpressionVisitorList.get(i);
+            String order = orderByAttributeBuilders.get(i).getOrder().name();
+            if (visitor.getStreamVarCount() == 0) {
+                compiledOrderByJSON.append(visitor.getCompiledCondition())
+                        .append((order.equalsIgnoreCase("ASC")) ? ":1" : ":-1")
+                        .append((orderByExpressionVisitorList.indexOf(visitor) ==
+                                (orderByExpressionVisitorList.size() - 1)) ? '}' : ',');
+            } else {
+                throw new MongoTableException("The MongoDB Event table does not support order by using " +
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]:
```suggestion
                throw new MongoTableException("The MongoDB Event table does not support 'order by' clause with " +
```
Make similar changes in other logs

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoExpressionVisitor.java [hunk 14]
```diff
@@ -66,10 +68,42 @@ public String getCompiledCondition() {
         return compiledCondition;
     }
 
+    public String getHavingCompiledCondition() {
+        if (isNullCheck) {
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoExpressionVisitor.java]: By having this check it will not run 'is null' in on condition also. Initiate the MongoExpressionVisitor with a flag to see if you are processing on condition/having and use that to throw an error. Remove all other changes it is unnecessary

File: component/src/test/java/io/siddhi/extension/store/mongodb/QueryableMongoTableTest.java [hunk 15]
```diff
@@ -0,0 +1,1126 @@
...
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class QueryableMongoTableTest {
+
+    private static final Logger log = Logger.getLogger(JoinMongoTableTest.class);
+
+    private static String uri = MongoTableTestUtils.resolveBaseUri();
+    private AtomicInteger eventCount = new AtomicInteger(0);
+    private int waitTime = 50;
```
Comment (niveathika) [inline component/src/test/java/io/siddhi/extension/store/mongodb/QueryableMongoTableTest.java]: Increase wait time or else it can intermittently fail

File: component/src/test/java/io/siddhi/extension/store/mongodb/QueryableMongoTableTest.java [hunk 16]
```diff
@@ -0,0 +1,1126 @@
...
+        SiddhiAppRuntime siddhiAppRuntime = siddhiManager.createSiddhiAppRuntime(streams + query);
+        siddhiAppRuntime.addCallback("query2", new QueryCallback() {
+            @Override
+            public void receive(long timeStamp, Event[] inEvents, Event[] removeEvents) {
+                if (inEvents != null) {
+                    for (Event event : inEvents) {
+                        eventCount.incrementAndGet();
+                        switch (eventCount.intValue()) {
+                            case 1:
+                                Assert.assertEquals(new Object[]{"WSO2", 100, 10, 100}, event.getData());
```
Comment (niveathika) [inline component/src/test/java/io/siddhi/extension/store/mongodb/QueryableMongoTableTest.java]: Use SiddhiTetHelper util class to assert, it's easier and cleaner

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java [hunk 17]
```diff
@@ -623,4 +626,262 @@ protected void destroy() {
...
+            aggregateList.add(orderBy);
+            MongoTableUtils.printDebugLogs("OrderBy query : ", orderBy.toJson());
+        }
+        if (offsetValue != null) {
+            Document offsetFilter = Document.parse("{ \'$skip\' :" + offsetValue + "}");
+            aggregateList.add(offsetFilter);
+            MongoTableUtils.printDebugLogs("Offset query : ", offsetFilter.toJson());
+        }
+        if (limitValue != null) {
+            Document limitFilter = Document.parse("{ \'$limit\' :" + limitValue + "}");
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: No need to escape single quotes,

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+    private Stack<String> conditionOperands;
+    private int streamVarCount;
+    private int constantCount;
+    private String[] supportedFunctions = {"sum", "avg", "min", "max", "count"};
+    private boolean isAttributeFunctionUsed;
+    private boolean isCountFunction;
+    private boolean isNullCheck;
+
+    public MongoSelectExpressionVisitor() {
+        this.conditionOperands = new Stack<>();
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Never initialize generalised objects
```suggestion
        this.conditionOperands = new Stack<String>();
```

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java [hunk 19]
```diff
@@ -0,0 +1,307 @@
...
+        return this.constantCount;
+    }
+
+    public boolean checkFunctionsUsed() {
+        return this.isAttributeFunctionUsed;
+    }
+
+    @Override
+    public void beginVisitConstant(Object value, Attribute.Type type) {
+        conditionOperands.push("{\'$literal\':" + value + "}");
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Move this to constants, saw the same string multiple times. Move ass MongoDB related query strings to contants

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+
+    @Override
+    public void beginVisitMath(MathOperator mathOperator) {
+    }
+
+    @Override
+    public void endVisitMath(MathOperator mathOperator) {
+        String rightOperand = this.conditionOperands.pop();
+        String leftOperand = this.conditionOperands.pop();
+        conditionOperands.push("{$" + mathOperator.name().toLowerCase() + ":" + "[" + leftOperand + "," +
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Use constants similar to https://github.com/siddhi-io/siddhi-store-mongodb/blob/master/component/src/main/java/io/siddhi/extension/store/mongodb/util/MongoTableConstants.java#L51

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java [hunk 21]
```diff
@@ -0,0 +1,307 @@
...
+
+    @Override
+    public void beginVisitOr() {
+    }
+
+    @Override
+    public void endVisitOr() {
+        String rightOperand = this.conditionOperands.pop();
+        String leftOperand = this.conditionOperands.pop();
+        conditionOperands.push("{$or:" + "[" + leftOperand + "," + rightOperand + "]}");
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Reuse https://github.com/siddhi-io/siddhi-store-mongodb/blob/master/component/src/main/java/io/siddhi/extension/store/mongodb/util/MongoTableConstants.java#L52

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+
+    @Override
+    public void beginVisitAnd() {
+    }
+
+    @Override
+    public void endVisitAnd() {
+        String rightOperand = this.conditionOperands.pop();
+        String leftOperand = this.conditionOperands.pop();
+        conditionOperands.push("{$and:" + "[" + leftOperand + "," + rightOperand + "]}");
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Same

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+            case NOT_EQUAL:
+                compareFilter = "$ne";
+                break;
+            default:
+                throw new MongoTableException("MongoDB Event Table found unknown operator '" + operator + "' for " +
+                        "COMPARE operation.");
+        }
+        String rightOperand = this.conditionOperands.pop();
+        String leftOperand = this.conditionOperands.pop();
+        conditionOperands.push("{ $cond: { if: { " + compareFilter + ": [" + leftOperand + "," + rightOperand + " ] }, then: true," +
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Reuse https://github.com/siddhi-io/siddhi-store-mongodb/blob/master/component/src/main/java/io/siddhi/extension/store/mongodb/util/MongoTableConstants.java#L54 also do the same check as in https://github.com/siddhi-io/siddhi-store-mongodb/blob/master/component/src/main/java/io/siddhi/extension/store/mongodb/MongoExpressionVisitor.java#L215  . Applicable to other methods

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java [hunk 24]
```diff
@@ -0,0 +1,307 @@
...
+    public void beginVisitCompareRightOperand(Compare.Operator operator) {
+    }
+
+    @Override
+    public void endVisitCompareRightOperand(Compare.Operator operator) {
+    }
+
+    @Override
+    public void beginVisitNot() {
+        throw new MongoTableException("MongoDB Event Table does not support NOT function.");
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Not is supported in MongoDB, https://github.com/siddhi-io/siddhi-store-mongodb/blob/master/component/src/main/java/io/siddhi/extension/store/mongodb/util/MongoTableConstants.java#L53

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java [hunk 25]
```diff
@@ -0,0 +1,307 @@
...
+    }
+
+    @Override
+    public void beginVisitCompare(Compare.Operator operator) {
+    }
+
+    @Override
+    public void endVisitCompare(Compare.Operator operator) {
+        String compareFilter;
+        switch (operator) {
```
Comment (niveathika) [inline component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: The checks and logic seem to be same as in MongoExpressionVisitor.endVisitCompare() except the wrapping of `{ $cond: { if: { }}}` see if you can reuse the method by moving them to util class.

**Review by niveathika — CHANGES_REQUESTED** [review]

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 2 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/461175f945869126b79d0590f0d26bd9a47727aa

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 10 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/94fba768238670966474ed6ac8a757455790895b

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 11 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/94fba768238670966474ed6ac8a757455790895b

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 17 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/94fba768238670966474ed6ac8a757455790895b

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java (hunk 18 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/94fba768238670966474ed6ac8a757455790895b

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java (hunk 19 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Fixed in 
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/94fba768238670966474ed6ac8a757455790895b

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java (hunk 20 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Fixed in 
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/94fba768238670966474ed6ac8a757455790895b

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java (hunk 21 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Fixed in 
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/94fba768238670966474ed6ac8a757455790895b

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java (hunk 22 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Fixed in 
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/94fba768238670966474ed6ac8a757455790895b

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 4 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/a270f77b9c9636c73cf908db7b64f418359d589f

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 5 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/a270f77b9c9636c73cf908db7b64f418359d589f

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 6 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java:730]: Test case added
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/a270f77b9c9636c73cf908db7b64f418359d589f

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 8 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java:805]: Test case added
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/a270f77b9c9636c73cf908db7b64f418359d589f

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 9 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/a270f77b9c9636c73cf908db7b64f418359d589f

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 13 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/a270f77b9c9636c73cf908db7b64f418359d589f

File: component/src/test/java/io/siddhi/extension/store/mongodb/QueryableMongoTableTest.java (hunk 15 above)
Comment (PrabodDunuwila) [reply component/src/test/java/io/siddhi/extension/store/mongodb/QueryableMongoTableTest.java]: Set wait time as 100.
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/a270f77b9c9636c73cf908db7b64f418359d589f

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 7 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Fixed in 
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/001f253c197301e24e6323affe151c9dce39b97a

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java (hunk 24 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/19316e5642550ec5d4d154a11df69b7c14d8d726

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 6 above)
Comment (niveathika) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java:730]: Not selecting t.symbol as null, I meant selecting without as `as` keyword,
select t.name, t.symbol....

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 7 above)
Comment (niveathika) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Remove the redundant variables in the visitor too

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 8 above)
Comment (niveathika) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java:805]: Same as in https://github.com/siddhi-io/siddhi-store-mongodb/pull/56#discussion_r348927322

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java (hunk 25 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/19316e5642550ec5d4d154a11df69b7c14d8d726

File: .gitignore (hunk 1 above)
Comment (PrabodDunuwila) [reply .gitignore]: Fix in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/fe9f20f192c0d2d19fb82c0a974aa5a6f478f860

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 6 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java:730]: Test case added
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/ab3c745d9931da13248bc248b8d4975c09951a52

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 8 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java:805]: Test case added
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/ab3c745d9931da13248bc248b8d4975c09951a52

File: component/src/test/java/io/siddhi/extension/store/mongodb/QueryableMongoTableTest.java (hunk 16 above)
Comment (PrabodDunuwila) [reply component/src/test/java/io/siddhi/extension/store/mongodb/QueryableMongoTableTest.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/dcb8942e73f14f6524648543b041073456fa55ef

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 12 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java:828]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/b10f2f4ef773fda6881b343866ea2f94aa210104

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoExpressionVisitor.java (hunk 14 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoExpressionVisitor.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/7593576ed5a33e7a78ea8db4a24c440b6fccf79e

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java (hunk 23 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: 1. Cannot reuse since their order of operands & operator is different. Created a new variable.
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/94fba768238670966474ed6ac8a757455790895b#

Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: 2. In SelectionExpressionVisitor do I need to do the conditional checks? Current implementation can support for multiple expressions as in mongodb.

Comment (niveathika) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoSelectExpressionVisitor.java]: Then it's fine. Let's not add that

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoExpressionVisitor.java (hunk 14 above)
Comment (niveathika) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoExpressionVisitor.java]: No need to wait till the getCompiledCondition() If the isHavingClause is true as soon as the beginVisistIsNull is hit, throw the error else continue to process

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java (hunk 3 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoDBEventTable.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/fd69ed50bdc97bea25664c44ea0eb562fd66bc08

File: component/src/main/java/io/siddhi/extension/store/mongodb/MongoExpressionVisitor.java (hunk 14 above)
Comment (PrabodDunuwila) [reply component/src/main/java/io/siddhi/extension/store/mongodb/MongoExpressionVisitor.java]: Fixed in
https://github.com/siddhi-io/siddhi-store-mongodb/pull/56/commits/fd69ed50bdc97bea25664c44ea0eb562fd66bc08

**Review by niveathika — APPROVED** [review]
